### PR TITLE
Permissioned contract deployment

### PR DESCRIPTION
**Pull request by Szegoo**
Closes: #3196

### CODE REVIEW

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5247525

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 2/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5247543

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 1/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5247542

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable-int
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5247849

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5247960

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5248114

**Review by athei — CHANGES_REQUESTED** [review]
This PR does not check the `InstantiateOrigin` when a contract instantiates another contract. Is this intended? No matter what the answer this has to be documented on the `Config` knob.

**Comment by Szegoo** [comment]
> This PR does not check the `InstantiateOrigin` when a contract instantiates another contract. Is this intended? 

Yes, this was intended, as I think it should probably be up to the contract logic to regulate other contract instantiation.

File: substrate/frame/contracts/src/lib.rs [hunk 1]
```diff
@@ -367,6 +367,14 @@ pub mod pallet {
 		#[pallet::constant]
 		type MaxDebugBufferLen: Get<u32>;
 
+		/// Origin allowed to upload code.
```
Comment (athei) [inline substrate/frame/contracts/src/lib.rs:370]: We need some explanation here that it is a safe default to allow everyone to upload code and permissioned code deployment is a special case.

File: substrate/frame/contracts/src/lib.rs [hunk 2]
```diff
@@ -367,6 +367,14 @@ pub mod pallet {
...
+		type UploadOrigin: EnsureOrigin<Self::RuntimeOrigin, Success = Self::AccountId>;
+
+		/// Origin allowed to instantiate code.
+		///
+		/// NOTE: This is not enforced when a contract instantiates another contract.
```
Comment (athei) [inline substrate/frame/contracts/src/lib.rs:375]:
```suggestion
		/// Origin allowed to instantiate code.
		///
		/// # Note
		///
		/// This is not enforced when a contract instantiates another contract. The [`UploadOrigin`] should
		/// make sure that no code is deployed that does unwanted instantiations.
```

**Review by athei — COMMENTED** [review]
Tests are insufficient. The instantiation extrinsics are not tested. You also need to define a special implementation for `EnsureOrigin` where you can change what it does dynamically. Have a look how we do that with `parameter_types!` for our `TestExtension`.  This will allow you to change what it does for your tests while leave it as-is for the existing tests.

**Comment by Szegoo** [comment]
> The instantiation extrinsics are not tested. 

It was actually tested, the instantiation origin was configured to allow any signed origin to deploy, and there was already a `root_cannot_instantiate` test. 

Based on your suggestion, I added more tests by dynamically restricting upload and instantiation, so it is better tested now.

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5406537

File: substrate/frame/contracts/src/lib.rs [hunk 3]
```diff
@@ -367,6 +367,23 @@ pub mod pallet {
...
+		///
+		/// By default, it is safe to set this to `EnsureSigned`, allowing anyone to upload contract
+		/// code.
+		type UploadOrigin: EnsureOrigin<Self::RuntimeOrigin, Success = Self::AccountId>;
+
+		/// Origin allowed to instantiate code.
+		///
+		/// # Note
+		///
+		/// This is not enforced when a contract instantiates another contract. The [`UploadOrigin`]
```
Comment (athei) [inline substrate/frame/contracts/src/lib.rs:380]:
```suggestion
		/// This is not enforced when a contract instantiates another contract. The [`Self::UploadOrigin`]
```

File: substrate/frame/contracts/src/tests.rs [hunk 4]
```diff
@@ -453,6 +454,58 @@ impl Contains<RuntimeCall> for TestFilter {
 	}
 }
 
+parameter_types! {
+	static UploadOrigin: <Test as frame_system::Config>::AccountId = ALICE;
+	static InstantiateOrigin: <Test as frame_system::Config>::AccountId = ALICE;
```
Comment (athei) [inline substrate/frame/contracts/src/tests.rs:459]:
```suggestion
	static UploadAccount: <Test as frame_system::Config>::AccountId = ALICE;
	static InstantiateAccount: <Test as frame_system::Config>::AccountId = ALICE;
```

**Review by athei — APPROVED** [review]

**Comment by pgherveou** [comment]
@Szegoo seeing your PR a bit late commented here
https://github.com/paritytech/polkadot-sdk/issues/3196#issuecomment-1976047906

Haven't look at the code here yet, but from the description it seems that this is something that can be done already. Can you double check and tell me what you think

**Comment by Szegoo** [comment]
This is more configurable. `dispatch_as` requires the origin to be root, while here the upload and instantiate origin could be set to anything(e.g. a collective).

File: substrate/frame/contracts/src/tests.rs [hunk 5]
```diff
@@ -453,6 +454,60 @@ impl Contains<RuntimeCall> for TestFilter {
...
+	}
+
+	#[cfg(feature = "runtime-benchmarks")]
+	fn try_successful_origin() -> Result<T::RuntimeOrigin, ()> {
+		Ok(T::RuntimeOrigin::from(frame_system::RawOrigin::Signed(
+			InstantiateAccount::get().into(),
+		)))
+	}
+}
+
```
Comment (pgherveou) [inline substrate/frame/contracts/src/tests.rs:510]: Maybe we can just  use a single type here for both Upload & Instantiate Origin

```suggestion
parameter_types! {
	pub static UploadAccount: Option<<Test as frame_system::Config>::AccountId> = None;
	pub static InstantiateAccount: Option<<Test as frame_system::Config>::AccountId> = None;
}

pub struct EnsureAccount<T, A>(sp_std::marker::PhantomData<(T, A)>);
impl<T: Config, A: sp_core::Get<Option<crate::AccountIdOf<T>>>>
	EnsureOrigin<<T as frame_system::Config>::RuntimeOrigin> for EnsureAccount<T, A>
where
	<T as frame_system::Config>::AccountId: From<AccountId32>,
{
	type Success = T::AccountId;

	fn try_origin(o: T::RuntimeOrigin) -> Result<Self::Success, T::RuntimeOrigin> {
		let who = <frame_system::EnsureSigned<_> as EnsureOrigin<_>>::try_origin(o.clone())?;
		if matches!(A::get(), Some(a) if who != a) {
			return Err(o)
		}

		Ok(who)
	}

	#[cfg(feature = "runtime-benchmarks")]
	fn try_successful_origin() -> Result<T::RuntimeOrigin, ()> {
		Err(())
	}
}
```

Then fix up 
```diff
-	type UploadOrigin = EnsureUploadAccount<Self>;
-	type InstantiateOrigin = EnsureInstantiateAccount<Self>;
+	type UploadOrigin = EnsureAccount<Self, UploadAccount>;
+	type InstantiateOrigin = EnsureAccount<Self, InstantiateAccount>;
 	type MaxDebugBufferLen = ConstU32<{ 2 * 1024 * 1024 }>;
 	type RuntimeHoldReason = RuntimeHoldReason;
 	type Migrations = crate::migration::codegen::BenchMigrations;
@@ -6007,7 +5980,7 @@ fn root_cannot_instantiate() {
 #[test]
 fn only_upload_origin_can_upload() {
 	let (wasm, _) = compile_module::<Test>("dummy").unwrap();
-	RestrictedUpload::set(true);
+	UploadAccount::set(Some(ALICE));
```

**Review by pgherveou — APPROVED** [review]

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5431858

**Review by athei — CHANGES_REQUESTED** [review]
> `dispatch_as` requires the origin to be root, while here the upload and instantiate origin could be set to anything(e.g. a collective).

Are you sure this is true? You should be able to check the origin in the filter. Zeitgeist is not doing this but you could do that, right?

**Comment by Szegoo** [comment]
> Are you sure this is true?

I was referring to:
https://github.com/paritytech/polkadot-sdk/blob/ec30d2f5da839a7fb3f34778013c63e5afc9e90f/substrate/frame/utility/src/lib.rs#L390

Not so sure about checking the origin in the filter though.

**Comment by athei** [comment]
Yes this is understood. Probably should have quoted this:

> This is more configurable.

Because you don't need to use `dispatch_as` as Zeitgeist does I think. You should be able to check the origin in the filter.

**Comment by pgherveou** [comment]
> > `dispatch_as` requires the origin to be root, while here the upload and instantiate origin could be set to anything(e.g. a collective).
> 
> Are you sure this is true? You should be able to check the origin in the filter. Zeitgeist is not doing this but you could do that, right?

I don't think you can the filter only give you access to the RuntimeCall

**Comment by pgherveou** [comment]
https://github.com/paritytech/polkadot-sdk/blob/c34ad77d0524add1ef7cec3de72df7874c67f8a9/substrate/frame/system/src/lib.rs?plain=1#L444-L455

**Review by athei — APPROVED** [review]
Yes you are right. The filter does not allow access to the origin.

**Pull request merged**